### PR TITLE
2020-02-06 GUARD-394 Products-Pulling-Error

### PR DESCRIPTION
**Pull request by fabulaspb**
**Summary**
During the product pulling BigCommerce's server can immediately close connection. If requested number of products per page is decreased on our side the issue disappears. Using V3 catalog API we can include in one response also all product's variations. Some customer's products has more than 250 variations. I guess the reason of such weird behavior from BigCommerce server is too big response. I added dynamic product page size changing in case of such error.

### CODE REVIEW

File: src/BigCommerceAccess/BigCommerceProductsServiceV3.cs [hunk 1]
```diff
@@ -29,9 +32,27 @@ public BigCommerceProductsServiceV3( WebRequestServices services ) : base( servi
...
+				{
+					var webEx = ex?.InnerException as WebException;
+
+					if ( webEx != null )
+					{
+						if ( webEx.Status == WebExceptionStatus.ConnectionClosed )
+						{
+							// gracefully decrease products page size
+							int newRequestMaxLimit = (int)Math.Floor( this.RequestMaxLimit / 2d );
+							i = (int)Math.Floor( ( ( i - 1 ) * this.RequestMaxLimit * 1.0 ) / newRequestMaxLimit ) + 1;
```
Comment (maxim-skuvault) [inline src/BigCommerceAccess/BigCommerceProductsServiceV3.cs]: This calculation is a little obscure. I tested with a RequestMaxLimit that's divisible by 2 and one that isn't and in both cases this  line could be simplified to `i = ( i - 1 ) * 2 - 1;` Can you think of a case when this won't work?

Comment (maxim-skuvault) [reply src/BigCommerceAccess/BigCommerceProductsServiceV3.cs]: In general it feels unclean to update the i in a `for` loop. But I guess if we restart the loop with the new limit then we'd lose the products already retrieved:(

Comment (fabulaspb) [reply src/BigCommerceAccess/BigCommerceProductsServiceV3.cs]: The idea is to decrease product page size each time when BigCommerce immediately closes connection due to huge amount of data to pass.
Let's imagine that such situation happened on page 3 with standard page size 250 items. In this case we need to drastically change current page size and calculate how many pages were already loaded and what value new page index should have. 
We already have successfully loaded 250 * 2 = 500 items. New page size is 250 / 2 = 125 items. So, new page index should be 500 / 125  + 1 = 5. This page index will be used next during retry attempt. Next loop cycle perhaps will be without issue and we'll request page index equals 6.
If we fail again with page size 125 on page 7 for example, we'll again decrease page size in half and we'll get next values: page size = 62, items loaded 6 * 125 = 750, new page index = 13, so data range will be [750, 812].

Comment (timjen3) [reply src/BigCommerceAccess/BigCommerceProductsServiceV3.cs]: This is clever. To piggy back on what Max was saying, you could change to using a while ( true ) loop instead and introduce a local pageNumber variable.

You could revert to the default page size after successfully getting a failed page:

Page 3 with page number 250 fails.
Retrieve page number 5 and 6 with page size 125.
Revert page size to 250 and page number to 6 / 2 + 1 = 4.

Not sure if it is worth the convolution but something to consider!

**Pull request merged**